compiler/AST/iterator.cpp

```diff
@@ -68,6 +69,267 @@ IteratorInfo::IteratorInfo() :
+  are not actually used in the program.
+
+* The absence of a parallel iterator is not an error. An error may
+  be issued later when+if this iterator is required for a parallel
+  computation.
+
+* The follower iterator is not sought at this time.  to seek it we would
+  need to know what the leader iterator yields. For that we may have to
+  resolve the **body** of the leader. Which we are not doing here.
+
+*/
+
+std::map<FnSymbol*,IteratorGroup*> iteratorGroups;
+
+IteratorGroup::IteratorGroup() :
+  serial(NULL), standalone(NULL), leader(NULL), follower(NULL) {}
+
+static bool isGroupable(FnSymbol* it) {
+  return it->hasFlag(FLAG_ITERATOR_FN) ||
+         it->hasFlag(FLAG_FN_RETURNS_ITERATOR);
+}
+
+// Return the IteratorGroup for the iterator 'it'.
+// If none, return NULL.
+IteratorGroup* getIteratorGroup(FnSymbol* it) {
+  if (! isGroupable(it))             return NULL;
+  std::map<FnSymbol*,IteratorGroup*>::iterator itig = iteratorGroups.find(it);
+  if (itig == iteratorGroups.end())  return NULL;
+  else                               return itig->second;
+}
+
+// Implements the rules listed under "Iterator Groups" in iterator.cpp.
+void resolveAlsoParallelIterators(FnSymbol* serial, Expr* call) {
+  if (! serial || ! isGroupable(serial))
+    return;  // not an iterator
+
+  IteratorGroup*& igroupRef = iteratorGroups[serial];
+
+  if (igroupRef != NULL)
+    return;  // already taken care of
+
+  for_formals(formal, serial)
+    if (formal->name == astrTag && formal->type == gLeaderTag->type)
+      // 'serial' is actually a parallel iterator. Since we did not come
+      // from a serial iterator, we will not update iterator groups.
+      return;
+
+  // This is first time we are seeing this serial iterator.
+  // Create an iterator group for it.
+  IteratorGroup* igroup = new IteratorGroup();
+  igroupRef = igroup;
+  igroup->serial = serial;
+
+  // ... and populate it. Use this representative call...
+  UnresolvedSymExpr* name = new UnresolvedSymExpr(serial->name);
+  CallExpr* repCall = new CallExpr(name);
+  call->getStmtExpr()->insertAfter(repCall);
+
+  for_formals(formal, serial)
+    repCall->insertAtTail(
+      new NamedExpr(formal->name, symOrParamExpr(formal)) );
+
+  // ... to look for the standalone then leader iterators.
+  NamedExpr* tag = new NamedExpr("tag", new SymExpr(gStandaloneTag));
+  repCall->insertAtTail(tag);
+
+  if (FnSymbol* SA = tryResolveCall(repCall)) {
+    // If !isGroupable(SA), we will report an error about it
+    // when about to use it in a forall loop.
+    //INT_ASSERT(isGroupable(SA));
+    igroup->standalone = SA;
+    iteratorGroups[SA] = igroup;
+    repCall->baseExpr->replace(name);     // reset repCall
+    repCall->insertAtTail(tag->remove()); // 'tag' may have been reordered
+  }
+
+  tag->actual->replace(new SymExpr(gLeaderTag));
+  if (FnSymbol* L = tryResolveCall(repCall)) {
+    // !isGroupable(L) might happen - see !isGroupable(SA) above.
```

Comment:
I'm not really understanding this isGroupable situation. Is it just the case that the call we tried to resolve doesn't necessarily resolve to an iterator? That seems like an obvious potentiality. So maybe the comments here should not say anything about isGroupable but instead say something like `// Note that L might not be an iterator at all. Errors in that event are raised later`.